### PR TITLE
feat: nicer http body reading, PoC for data_info field

### PR DESCRIPTION
**Pull request by untitaker**

### CODE REVIEW

**Comment by untitaker** [comment]
@mitsuhiko relevant changes are in `_wsgi.py` and `flask.py`

File: sentry_sdk/integrations/flask.py [hunk 1]
```diff
@@ -73,30 +78,77 @@ def _before_request(*args, **kwargs):
...
     }
 
-
-def _get_user_info():
+    scope.request = request_info
+    # if this crashes we at least have the rest of the request already set
+    _set_request_body(request_info, scope)
+
+
+def _set_request_body(request_info, scope):
```
Comment (untitaker) [inline sentry_sdk/integrations/flask.py]: I will have a serious problem with code reuse for the django integration here, that's why I am considering doing it on the wsgi level. But alas Django allegedly sucks for adding custom wsgi middleware anyway, so I'm not sure

**Comment by mitsuhiko** [comment]
Instead of doing all this heavy lifting every time we start a request I would propose we register an event processor on the scope instead. It will be invoked before the first exception is thrown in the scope and can then pull data as needed.  That also has a smaller chance of screwing up the request state and I think we can get away without doing wrapping on the request stream.

**Comment by mitsuhiko** [comment]
So i mean we have something like this:

```python
@Hub.get_current().add_event_processor
def make_flask_processor():
    data = request.form
    def processor(event):
        event['http']... = data
    return processor
```

**Comment by untitaker** [comment]
@mitsuhiko we can do this but we still need to wrap the data stream like in `wsgi.py`#

>I think we can get away without doing wrapping on the request stream.

what do you mean by this?

**Comment by mitsuhiko** [comment]
I think consuming json/form data by the time someone emits an event is fair. We did it in the past and it did not cause too many concerns.

**Comment by untitaker** [comment]
yes, but if somebody emits an event and the request data is neither json nor form, we would have to access `request.data` which might read way too much into memory. It might also miss data if somebody manually read from `request.stream`. That's my concern.

**Comment by untitaker** [comment]
If I understand correctly you propose accessing `form`, `json`, `data` in that order or similar, and use normal string truncation, right?

**Comment by mitsuhiko** [comment]
I'm not 100% sure. The thing is that reading from request.stream has side effects obviously so there is a good chance that we will just not do it at all in the Python integrations. So it's just `request.form` and `request.get_json()` in Flask.

For what it's worth i'm not fully opposed to the solution proposed in the PR right now. I just feel like we might have to put reading from the stream without mimetype check behind an option and then i'm not sure how many people would switch it on.

**Comment by untitaker** [comment]
Note though that the prefix/peek directly read from the stream is not used if it turns out that the data was form or json. In those cases the entire stream is still read and parsed through Flask/Werkzeug

**Comment by untitaker** [comment]
closing in favor of #11

**Pull request closed without merging**